From 02341853dbad27c70351f469363b0f80f64f6ac0 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Tue, 12 Sep 2017 09:32:13 +0200
Subject: [PATCH 1/1] efi_loader: efi_check_events should invoke notify
 function

If the event is not in the signaled state the notify function should be
invoked in efi_check_events (as required by the UEFI 2.7 spec).

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_boottime.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/lib/efi_loader/efi_boottime.c b/lib/efi_loader/efi_boottime.c
index 153a106ca1..b83676b343 100644
--- a/lib/efi_loader/efi_boottime.c
+++ b/lib/efi_loader/efi_boottime.c
@@ -479,8 +479,13 @@ static efi_status_t EFIAPI efi_check_event(struct efi_event *event)
 			continue;
 		if (!event->type || event->type & EVT_NOTIFY_SIGNAL)
 			break;
-		if (event->signaled)
+		if (!event->signaled && event->notify_function)
+			EFI_CALL_VOID(event->notify_function(event,
+						     event->notify_context));
+		if (event->signaled) {
+			event->signaled = 0;
 			return EFI_EXIT(EFI_SUCCESS);
+		}
 		return EFI_EXIT(EFI_NOT_READY);
 	}
 	return EFI_EXIT(EFI_INVALID_PARAMETER);
-- 
2.11.0

